File: app/src/main/java/com/pauloavelar/inventory/view/ProductsActivity.java
package com.pauloavelar.inventory.view;

import android.app.AlertDialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.support.v4.app.NavUtils;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.Menu;
import android.view.MenuItem;

import com.pauloavelar.inventory.R;
import com.pauloavelar.inventory.dao.ProductDAO;

public class ProductsActivity extends AppCompatActivity implements DialogInterface.OnClickListener {

    private BroadcastReceiver mReceiver;
    private ProductAdapter mAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_products);

        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle(R.string.manage_products);
            getSupportActionBar().setDisplayShowHomeEnabled(true);
            getSupportActionBar().setDefaultDisplayHomeAsUpEnabled(true);
        }

        RecyclerView list = (RecyclerView) findViewById(R.id.list_products);
        mAdapter = new ProductAdapter(new ProductAdapter.OnInteraction() {
            @Override
            public void onClickEdit(String product) {
                ProductDialogBuilder.showProductDialog(ProductsActivity.this, product);
            }

            @Override
            public void onClickDelete(String product) {
                ProductDAO.delete(ProductsActivity.this, product);
            }
        });
        list.setAdapter(mAdapter);
        list.addItemDecoration(new DividerItemDecoration(this));
        list.setLayoutManager(new LinearLayoutManager(this));
        loadProductList();

        mReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                loadProductList();
            }
        };

        LocalBroadcastManager.getInstance(this)
            .registerReceiver(mReceiver, new IntentFilter(ProductDAO.REFRESH_PRODUCT_LIST));
    }

    private void loadProductList() {
        mAdapter.setItems(ProductDAO.findAll(this));
    }

    @Override
    protected void onDestroy() {
        LocalBroadcastManager.getInstance(this).unregisterReceiver(mReceiver);
        super.onDestroy();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_products, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_add:
                ProductDialogBuilder.showProductDialog(this);
                return true;
            case R.id.action_clear_all:
                clickClearAll();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void clickClearAll() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(R.string.are_you_sure);
        builder.setMessage(R.string.delete_products_message);
        builder.setPositiveButton(R.string.yes_delete, this);
        builder.setNegativeButton(R.string.cancel, null);
        builder.show();
    }

    @Override
    public void onClick(DialogInterface dialogInterface, int which) {
        ProductDAO.deleteAll(this);
        mAdapter.clearAll();
    }

}

File: app/src/main/java/com/pauloavelar/inventory/ui/ProductsActivity.kt
package com.pauloavelar.inventory.ui

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.pauloavelar.inventory.R
import com.pauloavelar.inventory.databinding.ActivityProductsBinding
import com.pauloavelar.inventory.model.Product
import com.pauloavelar.inventory.ui.component.ListDividerDecoration
import com.pauloavelar.inventory.ui.component.ProductAdapter
import com.pauloavelar.inventory.ui.viewmodel.ProductsViewModel
import com.pauloavelar.inventory.utils.returnTrue
import org.koin.core.KoinComponent
import org.koin.core.inject

class ProductsActivity : AppCompatActivity(), KoinComponent {

    private lateinit var binding: ActivityProductsBinding
    private var productsAdapter: ProductAdapter? = null

    private val viewModel: ProductsViewModel by inject()
    private val dialogBuilder: DialogBuilder by inject()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityProductsBinding.inflate(layoutInflater)
        setContentView(binding.root)

        supportActionBar?.apply {
            setTitle(R.string.manage_products)
            setDisplayShowHomeEnabled(true)
        }

        productsAdapter = ProductAdapter(object : ProductAdapter.OnItemInteraction {
            override fun onClickEdit(product: Product) {
                dialogBuilder.showProductDialog(this@ProductsActivity, product)
            }

            override fun onClickDelete(product: Product) = viewModel.deleteProduct(product)
        })

        findViewById<RecyclerView>(R.id.list_products).apply {
            adapter = productsAdapter
            addItemDecoration(ListDividerDecoration(this@ProductsActivity))
            layoutManager = LinearLayoutManager(this@ProductsActivity)
        }
    }

    override fun onResume() {
        super.onResume()

        viewModel.getProducts().observe(this) {
            productsAdapter!!.setItems(it)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_products, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem) = when (item.itemId) {
        R.id.action_add -> returnTrue {
            dialogBuilder.showProductDialog(this@ProductsActivity)
        }
        R.id.action_clear_all -> returnTrue {
            dialogBuilder.showClearAllProductsDialog(this@ProductsActivity)
        }
        else -> {
            super.onOptionsItemSelected(item)
        }
    }

}
